### PR TITLE
fix: add and remove partition transform on same column failed when use v1 metadata

### PR DESCRIPTION
**Pull request by vinson0526**
When use v1 metadata, we cannot add  a new partition that previously existed.
Further more, exception will thrown when we update partition spec after dropped partition transform more than once on same column.

I use iceberg master branch and spark 3.0.2.
we can reproduce follow these step.
```
create table test(create_time timestamp) using iceberg;
alter table test add partition field years(create_time);
alter table test drop partition field years(create_time);
alter table test add partition field months(create_time);
alter table test drop partition field months(create_time);
alter table test add partition field days(create_time);
```
and exception thrown
```
Multiple entries with same key: (1, void)=1001: create_time_month: void(1) and (1, void)=1000: create_time_year: void(1)
```

I try to fix it by this PR.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (pvary) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:228]: I am not an expert with this part of the code, but the comment states: `field IDs were not required for v1`
Could we end up with the same `field.name() + "_" + null` if the above happens?

Maybe it would be better to add a random generated UUID instead to avoid any chance of conflicts.

**Comment by pvary** [comment]
Started the tests. Let's see if we break something, or not 😄 

Left one question in review, but generally looks good to me

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:228]: we cannot use constant name for removed field, because partition spec field name must be unique.
Since removed field is not really used, we need to change it name to some that not conflict with new field.
UUID cannot ensure generate unique.
field.fieldId() is unique, because we generate it from 1000 and auto-increment for new field.
Maybe field.name() + field.fieldId() + UUID is better?

Comment (pvary) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:228]: I know wikipedia is not every time the source of truth, but [here](https://en.wikipedia.org/wiki/Universally_unique_identifier) it states:

> A universally unique identifier (UUID) is a 128-bit label used for information in computer systems.
> [..]
> When generated according to the standard methods, UUIDs are, for practical purposes, unique."

But reading again the comment in the code, if the fieldId is always autogenerated and never `null` then we are fine with your solution as well.

Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:228]: thank you for provide useful information about UUID.
and sorry about ut failed again. I think it will be work with the newest commit.

**Review by pvary — APPROVED** [review]

**Comment by pvary** [comment]
@marton-bod, @lcspinter: Could you please take a look as well?

Thanks,
Peter

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (marton-bod) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:295]: Just a thought: would it make sense to just do `!(field.transform() instanceof VoidTransform)` here? Then you wouldn't need the extra interface method

**Review by marton-bod — COMMENTED** [review]
Looks good to me, just one question

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:295]: yes, in this case, they are same. i add the extra interface method because void transform is a particular transform, this method maybe used in other place later.  if we don't want to add extra interface method, i can change it to `!(field.transform() instanceof VoidTransform)`.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:228]: Why do this instead of just renaming the field when a conflict is detected later? That seems like a better way to fix it to me.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:295]: Is this needed?

The map created by this method is only used in `addField` and `removeField`. My guess is that your intent is to avoid blocking multiple void transforms with the same source id because of multiple calls to remove. But the `apply` method works with the builder directly and so this map doesn't need to change if I'm reading the existing code correctly. What is the case where this causes a failure?

Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:295]: ImmutableMap.Builder disallow multiple entries with same key, and multiple void transforms with the same source id generate same key here. So this builder will throw an exception with msg such as "Multiple entries with same key:  (1, void)=1001: create_time_month: void(1) and (1, void)=1000: create_time_year: void(1)".

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:228]: We have two choice when a a conflict is detected:
1. to rename the existing void transform, that will significant add complexity to [the check method](https://github.com/apache/iceberg/blob/35691728df3b6930045271b1e5663c8516b52301/api/src/main/java/org/apache/iceberg/PartitionSpec.java#L339), just to find which void transform conflict with the current added transform.
2. to rename the new transform, that will either broke default partition name consistency, or change the name assigned by user

**Review by marton-bod — APPROVED** [review]

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:295]: In that case, I think we should use `Maps.newHashMap` instead of `ImmutableMap.Builder`. That way we don't need to change the `Transform` API to add `isVoid`.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:228]: I don't think we want to modify that check method. Instead, this update class should be modified to rename void fields if necessary. That isn't very difficult because we can track whether there are existing fields with the same name and rename any field that uses a void transform.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java]: Use Maps.newHashMap instead of ImmutableMap.Builder in newest commit. Please take a look again.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java]: The newest commit modified update class to rename void fields if necessary. Please take a look again.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:161]: Why not create a new `PartitionField` with its name set to `partitionName`? Then you wouldn't need to change as much, including the type of `adds`.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:232]: Nit: indentation is off. Should be 4 spaces / 2 indents from the start of `if`.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:365]: +1 for this approach to detect void transforms.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:231]: The logic here is still not quite what I was picturing, although I think this is a bit cleaner than before.

What I was thinking is that we can detect name collisions in `addField` by checking for an existing field with some name. If one exists, then instead of throwing an exception we could check if it is a void transform and add a rename if it is. Something like this:

```java
  public BaseUpdatePartitionSpec addField(String name, Term term) {
    PartitionField existingField = nameToField.get(name);
    if (existingField != null) {
      if (isVoidTransform(existingField)) {
        renameField(name, name + "_" + existingField.fieldId());
      } else {
        throw new IllegalArgumentException(String.format("Cannot add duplicate partition field name: %s", name));
      }
    }
    ...
  }
```

Would that not work?

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:161]: If `name` is null, we generate default name for it at the end of `apply()` method. We need all name of added field before detect the conflict. So i change the type of `adds` to save generated name , avoid generate it twice.
if i don't change the type of `adds`, i think there are four other options:
1. just generate name twice
2. add a set method `setName(String name)` to `PartitionField`
3. generate newField again with generated name: just like
```java
if (newField.name() == null) {
    String partitionName = PartitionSpecVisitor.visit(schema, newField, PartitionNameGenerator.INSTANCE);
    newField = new PartitionField(
        newField.sourceId(), newField.fieldId(), partitionName, newField.transform());
}

adds.add(newField);
```
4. write a new static `visit` method in `PartitionSpecVisitor`, the signature is like:
```java
static <R> R visit(Schema schema, int sourceId, Transform<?, ?> transform, PartitionSpecVisitor<R> visitor)
```
Which do you think is better?

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:232]: fix it in next commit

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java]: This is better than my code. Already move detect code to `addField` and `renameField` in newest commit.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:171]: The `PartitionField` is created in this method, as is `partitionName`. I think that by reordering a couple of statements here, you can use the correct name on the field from the start and avoid more changes.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:217]: Why use `UUID.randomUUID()` instead of the partition field ID? I think it makes more sense to use `existingField.fieldId()`.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:161]: Here as well, I'd prefer not to use a UUID. This should be able to use the existing field's ID instead.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:171]: Regenerate `PartitionField` with `partitionName ` if `name` is null.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:217]: change to `existingField.fieldId()`

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java
Comment (vinson0526) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:161]: use the existing field's ID instead.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks for the update, @vinson0526! I kicked off the tests and will merge once they pass. The changes look great!

**Comment by pvary** [comment]
Since every test passed and got +1 from @rdblue, merged the change.

Thanks for the fix @vinson0526 and @rdblue, @marton-bod for the review!

**Pull request merged**
